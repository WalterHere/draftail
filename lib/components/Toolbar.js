import PropTypes from 'prop-types';
import React from 'react';

import DraftUtils from '../api/DraftUtils';

import ToolbarButton from '../components/ToolbarButton';
import ToolbarGroup from '../components/ToolbarGroup';

import {
    BR_TYPE,
    ENTITY_TYPE,
    UNDO_TYPE,
    REDO_TYPE,
    LABELS,
    DESCRIPTIONS,
} from '../api/constants';
import behavior from '../api/behavior';

const getButtonLabel = (type, icon, label = icon ? null : LABELS[type]) =>
    label;

const getButtonTitle = (type, description = DESCRIPTIONS[type]) => {
    const hasShortcut = behavior.hasKeyboardShortcut(type);
    let title = description;

    if (hasShortcut) {
        const desc = description ? `${description}\n` : '';
        title = `${desc}${behavior.getKeyboardShortcut(type)}`;
    }

    return title;
};

const Toolbar = ({
    editorState,
    blockTypes,
    inlineStyles,
    enableHorizontalRule,
    enableLineBreak,
    showUndoControl,
    showRedoControl,
    entityTypes,
    toggleBlockType,
    toggleInlineStyle,
    addHR,
    addBR,
    onUndoRedo,
    onRequestSource,
}) => (
    <div className="Draftail-Toolbar" role="toolbar">
        <ToolbarGroup>
            {inlineStyles.map(style => (
                <ToolbarButton
                    key={style.type}
                    name={style.type}
                    active={DraftUtils.hasCurrentInlineStyle(
                        editorState,
                        style.type,
                    )}
                    label={getButtonLabel(style.type, style.icon, style.label)}
                    title={getButtonTitle(style.type, style.description)}
                    icon={style.icon}
                    onClick={toggleInlineStyle.bind(null, style.type)}
                />
            ))}
        </ToolbarGroup>

        <ToolbarGroup>
            {blockTypes.map(block => (
                <ToolbarButton
                    key={block.type}
                    name={block.type}
                    active={DraftUtils.isSelectedBlockType(
                        editorState,
                        block.type,
                    )}
                    label={getButtonLabel(block.type, block.icon, block.label)}
                    title={getButtonTitle(block.type, block.description)}
                    icon={block.icon}
                    onClick={toggleBlockType.bind(null, block.type)}
                />
            ))}
        </ToolbarGroup>

        <ToolbarGroup>
            {enableHorizontalRule ? (
                <ToolbarButton
                    name={ENTITY_TYPE.HORIZONTAL_RULE}
                    onClick={addHR}
                    label={getButtonLabel(
                        ENTITY_TYPE.HORIZONTAL_RULE,
                        enableHorizontalRule.icon,
                        enableHorizontalRule.label,
                    )}
                    title={getButtonTitle(
                        ENTITY_TYPE.HORIZONTAL_RULE,
                        enableHorizontalRule.description,
                    )}
                    icon={enableHorizontalRule.icon}
                />
            ) : null}

            {enableLineBreak ? (
                <ToolbarButton
                    name={BR_TYPE}
                    onClick={addBR}
                    label={getButtonLabel(
                        BR_TYPE,
                        enableLineBreak.icon,
                        enableLineBreak.label,
                    )}
                    title={getButtonTitle(BR_TYPE, enableLineBreak.description)}
                    icon={enableLineBreak.icon}
                />
            ) : null}
        </ToolbarGroup>

        <ToolbarGroup>
            {entityTypes.map(entity => (
                <ToolbarButton
                    key={entity.type}
                    name={entity.type}
                    onClick={onRequestSource.bind(null, entity.type)}
                    label={getButtonLabel(
                        entity.type,
                        entity.icon,
                        entity.label,
                    )}
                    title={getButtonTitle(entity.type, entity.description)}
                    icon={entity.icon}
                />
            ))}
        </ToolbarGroup>

        <ToolbarGroup>
            {showUndoControl ? (
                <ToolbarButton
                    name={UNDO_TYPE}
                    onClick={onUndoRedo.bind(null, UNDO_TYPE)}
                    label={getButtonLabel(
                        UNDO_TYPE,
                        showUndoControl.icon,
                        showUndoControl.label,
                    )}
                    title={getButtonTitle(
                        UNDO_TYPE,
                        showUndoControl.description,
                    )}
                />
            ) : null}

            {showRedoControl ? (
                <ToolbarButton
                    name={REDO_TYPE}
                    onClick={onUndoRedo.bind(null, REDO_TYPE)}
                    label={getButtonLabel(
                        REDO_TYPE,
                        showRedoControl.icon,
                        showRedoControl.label,
                    )}
                    title={getButtonTitle(
                        REDO_TYPE,
                        showRedoControl.description,
                    )}
                />
            ) : null}
        </ToolbarGroup>
    </div>
);

Toolbar.propTypes = {
    editorState: PropTypes.object.isRequired,
    enableHorizontalRule: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.shape({
            // Describes the control in the editor UI, concisely.
            label: PropTypes.string,
            // Describes the control in the editor UI.
            description: PropTypes.string,
            // Represents the control in the editor UI.
            icon: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.arrayOf(PropTypes.string),
                PropTypes.node,
            ]),
        }),
    ]).isRequired,
    enableLineBreak: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.shape({
            // Describes the control in the editor UI, concisely.
            label: PropTypes.string,
            // Describes the control in the editor UI.
            description: PropTypes.string,
            // Represents the control in the editor UI.
            icon: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.arrayOf(PropTypes.string),
                PropTypes.node,
            ]),
        }),
    ]).isRequired,
    showUndoControl: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.shape({
            // Describes the control in the editor UI, concisely.
            label: PropTypes.string,
            // Describes the control in the editor UI.
            description: PropTypes.string,
            // Represents the control in the editor UI.
            icon: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.arrayOf(PropTypes.string),
                PropTypes.node,
            ]),
        }),
    ]).isRequired,
    showRedoControl: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.shape({
            // Describes the control in the editor UI, concisely.
            label: PropTypes.string,
            // Describes the control in the editor UI.
            description: PropTypes.string,
            // Represents the control in the editor UI.
            icon: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.arrayOf(PropTypes.string),
                PropTypes.node,
            ]),
        }),
    ]).isRequired,
    entityTypes: PropTypes.array.isRequired,
    blockTypes: PropTypes.array.isRequired,
    inlineStyles: PropTypes.array.isRequired,
    toggleBlockType: PropTypes.func.isRequired,
    toggleInlineStyle: PropTypes.func.isRequired,
    addHR: PropTypes.func.isRequired,
    addBR: PropTypes.func.isRequired,
    onUndoRedo: PropTypes.func.isRequired,
    onRequestSource: PropTypes.func.isRequired,
};

export default Toolbar;
