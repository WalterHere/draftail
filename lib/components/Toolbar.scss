.#{$DRAFTAIL}Toolbar {
    position: relative;
    padding: $controls-spacing;
    // Offset the button's top margin to form rows.
    padding-top: 0;
    background-color: $draftail-editor-chrome;
    color: $draftail-editor-chrome-text;
    border-radius: $draftail-toolbar-radius;
    border-bottom: $draftail-toolbar-border-bottom;

    .#{$DRAFTAIL}Editor--focus & {
        position: sticky;
        top: 0;
        z-index: $draftail-toolbar-z-index;
    }
}
