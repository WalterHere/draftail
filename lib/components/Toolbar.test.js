import React from 'react';
import { shallow } from 'enzyme';
import { EditorState } from 'draft-js';

import Toolbar from '../components/Toolbar';

const mockProps = {
    editorState: EditorState.createEmpty(),
    enableHorizontalRule: false,
    enableLineBreak: false,
    showUndoControl: false,
    showRedoControl: false,
    entityTypes: [],
    blockTypes: [],
    inlineStyles: [],
    toggleBlockType: () => {},
    toggleInlineStyle: () => {},
    addHR: () => {},
    addBR: () => {},
    onUndoRedo: () => {},
    onRequestSource: () => {},
};

describe('Toolbar', () => {
    it('empty', () => {
        expect(shallow(<Toolbar {...mockProps} />)).toMatchSnapshot();
    });

    describe('#enableHorizontalRule', () => {
        it('works', () => {
            expect(
                shallow(<Toolbar {...mockProps} enableHorizontalRule={true} />),
            ).toMatchSnapshot();
        });

        it('control overrides', () => {
            expect(
                shallow(
                    <Toolbar
                        {...mockProps}
                        enableHorizontalRule={{
                            icon: '#icon-hr',
                            label: 'HR',
                            description: 'Horizontal rule',
                        }}
                    />,
                ),
            ).toMatchSnapshot();
        });
    });

    describe('#enableLineBreak', () => {
        it('works', () => {
            expect(
                shallow(<Toolbar {...mockProps} enableLineBreak={true} />),
            ).toMatchSnapshot();
        });

        it('control overrides', () => {
            expect(
                shallow(
                    <Toolbar
                        {...mockProps}
                        enableLineBreak={{
                            icon: '#icon-br',
                            label: 'BR',
                            description: 'Soft line break',
                        }}
                    />,
                ),
            ).toMatchSnapshot();
        });
    });

    describe('#showUndoControl', () => {
        it('works', () => {
            expect(
                shallow(<Toolbar {...mockProps} showUndoControl={true} />),
            ).toMatchSnapshot();
        });

        it('control overrides', () => {
            expect(
                shallow(
                    <Toolbar
                        {...mockProps}
                        showUndoControl={{
                            icon: '#icon-undo',
                            label: 'Undo',
                            description: 'Undo last change',
                        }}
                    />,
                ),
            ).toMatchSnapshot();
        });
    });

    describe('#showRedoControl', () => {
        it('works', () => {
            expect(
                shallow(<Toolbar {...mockProps} showRedoControl={true} />),
            ).toMatchSnapshot();
        });

        it('control overrides', () => {
            expect(
                shallow(
                    <Toolbar
                        {...mockProps}
                        showRedoControl={{
                            icon: '#icon-redo',
                            label: 'Redo',
                            description: 'Redo last change',
                        }}
                    />,
                ),
            ).toMatchSnapshot();
        });
    });

    it('#entityTypes', () => {
        expect(
            shallow(
                <Toolbar
                    {...mockProps}
                    entityTypes={[
                        { type: 'TEST_1', label: 'Test 1' },
                        { type: 'TEST_2', label: 'Test 2' },
                        { type: 'TEST_3', label: 'Test 3' },
                    ]}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#blockTypes', () => {
        expect(
            shallow(
                <Toolbar
                    {...mockProps}
                    blockTypes={[
                        { type: 'TEST_1', label: 'Test 1' },
                        { type: 'TEST_2', label: 'Test 2' },
                        { type: 'TEST_3', label: 'Test 3' },
                    ]}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#inlineStyles', () => {
        expect(
            shallow(
                <Toolbar
                    {...mockProps}
                    inlineStyles={[
                        { type: 'TEST_1', label: 'Test 1' },
                        { type: 'TEST_2', label: 'Test 2' },
                        { type: 'TEST_3', label: 'Test 3' },
                    ]}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('button label default', () => {
        const wrapper = shallow(
            <Toolbar {...mockProps} inlineStyles={[{ type: 'BOLD' }]} />,
        );
        expect(wrapper.find('ToolbarButton').prop('label')).toBe('B');
    });

    it('button label with icon', () => {
        const wrapper = shallow(
            <Toolbar
                {...mockProps}
                inlineStyles={[{ type: 'BOLD', icon: '#icon-bold' }]}
            />,
        );
        expect(wrapper.find('ToolbarButton').prop('label')).toBe(null);
    });

    it('button label overrides', () => {
        const wrapper = shallow(
            <Toolbar
                {...mockProps}
                inlineStyles={[
                    {
                        type: 'BOLD',
                        label: 'Format as bold',
                    },
                ]}
            />,
        );
        expect(wrapper.find('ToolbarButton').prop('label')).toBe(
            'Format as bold',
        );
    });

    it('button titles with shortcut', () => {
        expect(
            shallow(
                <Toolbar
                    {...mockProps}
                    inlineStyles={[{ type: 'BOLD', description: null }]}
                />,
            )
                .find('ToolbarButton')
                .prop('title'),
        ).toBe('Ctrl + B');
    });

    it('button titles with shortcut and description', () => {
        expect(
            shallow(
                <Toolbar
                    {...mockProps}
                    inlineStyles={[{ type: 'BOLD', description: 'Bold' }]}
                />,
            )
                .find('ToolbarButton')
                .prop('title'),
        ).toBe('Bold\nCtrl + B');
    });
});
