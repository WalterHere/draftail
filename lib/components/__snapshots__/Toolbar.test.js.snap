// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar #blockTypes 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_1"
      label="Test 1"
      name="TEST_1"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_2"
      label="Test 2"
      name="TEST_2"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_3"
      label="Test 3"
      name="TEST_3"
      onClick={[Function]}
      title={null}
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #enableHorizontalRule control overrides 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon="#icon-hr"
      label="HR"
      name="HORIZONTAL_RULE"
      onClick={[Function]}
      title="Horizontal rule
- - -"
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #enableHorizontalRule works 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="―"
      name="HORIZONTAL_RULE"
      onClick={[Function]}
      title="Horizontal line
- - -"
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #enableLineBreak control overrides 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon="#icon-br"
      label="BR"
      name="BR"
      onClick={[Function]}
      title="Soft line break
Shift+Enter"
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #enableLineBreak works 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="↵"
      name="BR"
      onClick={[Function]}
      title="Line break
Shift+Enter"
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #entityTypes 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_1"
      label="Test 1"
      name="TEST_1"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_2"
      label="Test 2"
      name="TEST_2"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_3"
      label="Test 3"
      name="TEST_3"
      onClick={[Function]}
      title={null}
    />
  </ToolbarGroup>
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #inlineStyles 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_1"
      label="Test 1"
      name="TEST_1"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_2"
      label="Test 2"
      name="TEST_2"
      onClick={[Function]}
      title={null}
    />
    <ToolbarButton
      active={false}
      icon={null}
      key="TEST_3"
      label="Test 3"
      name="TEST_3"
      onClick={[Function]}
      title={null}
    />
  </ToolbarGroup>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;

exports[`Toolbar #showRedoControl control overrides 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="Redo"
      name="redo"
      onClick={[Function]}
      title="Redo last change
Ctrl+Shift+Z"
    />
  </ToolbarGroup>
</div>
`;

exports[`Toolbar #showRedoControl works 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="↻"
      name="redo"
      onClick={[Function]}
      title="Redo
Ctrl+Shift+Z"
    />
  </ToolbarGroup>
</div>
`;

exports[`Toolbar #showUndoControl control overrides 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="Undo"
      name="undo"
      onClick={[Function]}
      title="Undo last change
Ctrl+Z"
    />
  </ToolbarGroup>
</div>
`;

exports[`Toolbar #showUndoControl works 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarButton
      active={false}
      icon={null}
      label="↺"
      name="undo"
      onClick={[Function]}
      title="Undo
Ctrl+Z"
    />
  </ToolbarGroup>
</div>
`;

exports[`Toolbar empty 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
  <ToolbarGroup />
</div>
`;
